### PR TITLE
[GEOS-7032] You can see the content of the server by sending the right request to GeoServer

### PR DESCRIPTION
**Pull request by aaime**
This one prevents XML entity expansion attacks that can make GeoServer display the contents of arbitrary files in the file system, and should also prevent "XML bombs".

* https://osgeo-org.atlassian.net/browse/GEOS-7032

### CODE REVIEW

**Comment by aaime** [comment]
This one prevents XML entity expansion attacks that can make GeoServer display the contents of arbitrary files in the file system, and should also prevent "XML bombs".

This work follows up pr #193 by tapping some overlooked xml parsing paths and covering new ones that weren't there back at the time

**Comment by aaime** [comment]
Ah, please review carefully as this is a fix for a security issue, that we need to backport to the stable and maintenance branches.

File: src/extension/wps/wps-core/src/main/java/org/geoserver/wps/ppio/XMLPPIO.java [hunk 1]
```diff
@@ -76,6 +80,33 @@ public void encode( Object value, OutputStream os) throws Exception {
...
+     * @param configuration
+     * @return
+     */
+    protected Parser getParser(Configuration configuration) {
+        Parser parser = new Parser(configuration);
+        // spring injection would have been preferred, but it changes the order in which PPIO
+        // are listed, which in turn chances the behavior of WPS (e.g., default mimeType)
+        parser.setEntityResolver(getResolveProvider().getEntityResolver());
+        return parser;
+    }
```
Comment (jodygarnett) [inline src/extension/wps/wps-core/src/main/java/org/geoserver/wps/ppio/XMLPPIO.java]: So this makes sense, the fundamental change is supplying a EntryResolver to parsers before use.

File: src/main/src/main/java/org/geoserver/catalog/SLDHandler.java [hunk 2]
```diff
@@ -5,11 +5,28 @@
  */
```
Comment (jodygarnett) [inline src/main/src/main/java/org/geoserver/catalog/SLDHandler.java]: So general feedback that header needs to be updated (yeah I know it is not that important ...)

File: src/main/src/main/java/org/geoserver/util/NoExternalEntityResolver.java [hunk 3]
```diff
@@ -25,6 +25,7 @@
  */
 public class NoExternalEntityResolver implements EntityResolver {
 
+    public static final String ERROR_MESSAGE_BASE = "Entity resolution disallowed for ";
```
Comment (jodygarnett) [inline src/main/src/main/java/org/geoserver/util/NoExternalEntityResolver.java]: What does taking this out as a static final accomplish?

**Comment by jodygarnett** [comment]
So this pull request does what it sets out to and should be good to go.

As this is a follow up to #193 is there anyway we can keep this problem from being reintroduced in the future?

**Comment by smithkm** [comment]
Fixed the headers and the conflict here: https://github.com/smithkm/geoserver/tree/entity_resolve_fix

File: src/main/src/main/java/org/geoserver/util/NoExternalEntityResolver.java (hunk 3 above)
Comment (aaime) [inline src/main/src/main/java/org/geoserver/util/NoExternalEntityResolver.java]: Avoid repeating over and over the same string among differen tests, see all the tests checking for this string

**Comment by aaime** [comment]
Kevin, if it's good to go, feel free to close this one, push yours, and merge

**Comment by smithkm** [comment]
@aaime Yes that was the plan but I keep running into build issues when I try to do a local test. (three way merges always make me wary)  It seems to be working now and hopefully will be done (successfully) by the time I have the GWC 1.8-M0 release done.

**Pull request closed without merging**